kernel/src/scheduler.rs

```diff
@@ -79,15 +101,25 @@ pub trait Scheduler<C: Chip> {
     /// returns `false`, then `do_process` will exit with a `KernelPreemption`.
     ///
     /// `id` is the identifier of the currently active process.
-    unsafe fn continue_process(&self, _id: ProcessId, chip: &C) -> bool {
+    fn continue_process(&self, _id: ProcessId, chip: &C) -> bool {
         !(chip.has_pending_interrupts() || DeferredCall::has_tasks())
     }
 }
 
+/// Enum representing the possible work the kernel has to do
+#[derive(Clone, Copy)]
+pub enum KernelWorkType {
+    Interrupts,
+    DeferredCalls,
+}
+
 /// Enum representing the actions the scheduler can request in each call to
 /// `scheduler.next()`.
 #[derive(Copy, Clone)]
 pub enum SchedulingDecision {
+    /// Tell the kernel to do work, such as interrupts or deferred calls
+    KernelWork(KernelWorkType),
+
     /// Tell the kernel to run the specified process with the passed timeslice.
     /// If `None` is passed as a timeslice, the process will be run
     /// cooperatively.
```

Comment:
I think this is really the crux of the change. Should the scheduler choose between:
- sleep and a specific process, or
- sleep, a specific process, and kernel work?


Part of me thinks those are two separate tasks:
1. Should the kernel run a process or not?
2. If the kernel is going to run a process, which one?

The advantage of separating them is a pure process scheduler, like round robin, can exist in cooperation with something that is strategically choosing between kernel work and processes.

The disadvantage of separating them is that it would be hard or impossible to prioritize a particular process over kernel work, but prioritize kernel work over other processes.

Can we use traits to solve this somehow? Can we have one trait for just process scheduling, and one for kernel scheduling that extends the process schedule trait? I think that would be fairly straightforward, what I don't know about is how easy it would be to create a generic structure which implements what are right now default trait implementations.

One thing I'm worried about is the default trait implementations. That decoupling of implementation from the actual structure can lead to confusion. And even though we do it in the current scheduler trait we may not want to keep that design.

Reply:
Sorry for the late reply! I thought my comments were published as single line comments, but apparently I started a review. Again, I apologize for my late reply!